### PR TITLE
r/aws_lightsail_instance: Fix crash when encountering key-only tags

### PR DESCRIPTION
**Pull request by stefansundin**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

---

Fixes this crash:

```
panic: runtime error: invalid memory address or nil pointer dereference
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x44409fa]
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4:
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: goroutine 184 [running]:
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags.KeyValueTags.Map(0xc00082ff20, 0xc00082ff20)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags/key_value_tags.go:138 +0xda
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsLightsailInstanceRead(0xc000278e00, 0x51fa760, 0xc000180a00, 0xc000278e00, 0x0)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_lightsail_instance.go:216 +0x6eb
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000618200, 0xc000ff7c20, 0x51fa760, 0xc000180a00, 0xc00043cfb8, 0x0, 0x0)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:455 +0x119
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc00043cb00, 0x6e58080, 0xc000995440, 0xc000ff7b30, 0xc00043cb00, 0xc000995440, 0xc000a40a80)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:525 +0x3d8
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x635f2e0, 0xc00043cb00, 0x6e58080, 0xc000995440, 0xc000917ec0, 0x0, 0x6e58080, 0xc000995440, 0xc00098e380, 0x1bc)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3153 +0x217
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000d8000, 0x6e78cc0, 0xc000545e00, 0xc00092a200, 0xc0003c80f0, 0xa2400d0, 0x0, 0x0, 0x0)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:995 +0x460
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: google.golang.org/grpc.(*Server).handleStream(0xc0000d8000, 0x6e78cc0, 0xc000545e00, 0xc00092a200, 0x0)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:1275 +0xd97
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0003b0800, 0xc0000d8000, 0x6e78cc0, 0xc000545e00, 0xc00092a200)
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:710 +0xbb
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: created by google.golang.org/grpc.(*Server).serveStreams.func1
2020-03-10T19:07:52.845-0700 [DEBUG] plugin.terraform-provider-aws_v2.52.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/google.golang.org/grpc/server.go:708 +0xa1
```

Creating tags with an empty string as a value appears to correctly create these key-only tags.

Documentation:
- https://lightsail.aws.amazon.com/ls/docs/en_us/articles/amazon-lightsail-tags
- https://docs.aws.amazon.com/lightsail/2016-11-28/api-reference/API_Tag.html

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_lightsail_instance: Support key-only tags.
```

Output from acceptance testing:
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLightsailInstance_Tags'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLightsailInstance_Tags -timeout 120m
=== RUN   TestAccAWSLightsailInstance_Tags
=== PAUSE TestAccAWSLightsailInstance_Tags
=== CONT  TestAccAWSLightsailInstance_Tags
--- PASS: TestAccAWSLightsailInstance_Tags (80.99s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	82.636s
```

### CODE REVIEW

File: aws/internal/keyvaluetags/service_tags_gen.go [hunk 1]
```diff
@@ -1782,7 +1782,13 @@ func LightsailKeyValueTags(tags []*lightsail.Tag) KeyValueTags {
 	m := make(map[string]*string, len(tags))
 
 	for _, tag := range tags {
-		m[aws.StringValue(tag.Key)] = tag.Value
+		if tag.Value == nil {
```
Comment (ewbankkit) [inline aws/internal/keyvaluetags/service_tags_gen.go]: @stefansundin Thanks for this.
This source file is actually generated and will be recreated if `make gen` is run.
This behaviour may well be occurring for all relevant AWS services (I will do some spot tests) so should be fixed globally - changing

https://github.com/terraform-providers/terraform-provider-aws/blob/0286dba6e6f55c52863bdfa6662282a250a32a74/aws/internal/keyvaluetags/key_value_tags.go#L138

to

```go
 result[k] = aws.StringValue(v)
```

would be the more strategic fix.

**Comment by ewbankkit** [comment]
@stefansundin I opened an issue to track this.

Closes https://github.com/terraform-providers/terraform-provider-aws/issues/12368.

File: aws/internal/keyvaluetags/service_tags_gen.go (hunk 1 above)
Comment (stefansundin) [reply aws/internal/keyvaluetags/service_tags_gen.go]: Ok. Updated. Let me know if I should make more changes.

**Comment by stefansundin** [comment]
```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSLightsailInstance_Tags'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLightsailInstance_Tags -timeout 120m
=== RUN   TestAccAWSLightsailInstance_Tags
=== PAUSE TestAccAWSLightsailInstance_Tags
=== CONT  TestAccAWSLightsailInstance_Tags
--- PASS: TestAccAWSLightsailInstance_Tags (80.99s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	82.636s
```

**Comment by ewbankkit** [comment]
Verified acceptance test:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSLightsailInstance_Tags'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSLightsailInstance_Tags -timeout 120m
=== RUN   TestAccAWSLightsailInstance_Tags
=== PAUSE TestAccAWSLightsailInstance_Tags
=== CONT  TestAccAWSLightsailInstance_Tags
--- PASS: TestAccAWSLightsailInstance_Tags (70.07s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	70.117s
```

**Comment by bflad** [comment]
Hi @stefansundin 👋  Thank you for reporting this and sorry you ran into trouble here.

Rather than adjusting all of the generated service functions, we would prefer to fix the underlying `(keyvaluetags.KeyValueTags).Map()` panic as part some work we are doing to add support for Auto Scaling service tags: https://github.com/terraform-providers/terraform-provider-aws/pull/13868/files#diff-d536593b39c59c8af54d5145c3e6206eL180

If there are other services that are distinguishing between `nil`/`""` we can revisit this type of change on a per-service level.

Can you adjust this pull request to just being the `aws_lightsail_instance` testing and documentation updates and we will use this to verify Lightsail key-only tagging is working as expected? Thank you.

**Comment by stefansundin** [comment]
@bflad Do you want me to pull in master?

**Comment by bflad** [comment]
@stefansundin thanks so much! I was able to verify this updated test passes after #13868.

```
--- PASS: TestAccAWSLightsailInstance_Tags (75.80s)
```

We should expect to see the fix land in the next week or so.

**Pull request merged**